### PR TITLE
 ActiveModel::Attribute: elide dup for immutable types

### PR DESCRIPTION
**Pull request by byroot**
This is a redo of https://github.com/rails/rails/pull/53337
Which was reverted in https://github.com/rails/rails/pull/53347

`Attribute` is mostly immutable, but the deserialized value is memoized, and that value can potentially be mutable, and mutated.

When the value is immutable however, we can share instances.

Benchmark: https://gist.github.com/casperisfine/ae56bec1e7eecbff3a696b367e2bafa2

Before:

```
ruby 3.4.0dev (2024-12-12T09:30:43Z master 197a3efc75) +YJIT +PRISM [arm64-darwin23]
Warming up --------------------------------------
        ActiveRecord     4.664k i/100ms
Calculating -------------------------------------
        ActiveRecord     46.983k (± 1.0%) i/s   (21.28 μs/i) -    237.864k in   5.063292s
```

After:

```
ruby 3.4.0dev (2024-12-12T09:30:43Z master 197a3efc75) +YJIT +PRISM [arm64-darwin23]
Warming up --------------------------------------
        ActiveRecord     5.404k i/100ms
Calculating -------------------------------------
        ActiveRecord     53.502k (± 1.4%) i/s   (18.69 μs/i) -    270.200k in   5.051328s
```

### CODE REVIEW

File: activemodel/lib/active_model/attribute.rb [hunk 1]
```diff
@@ -96,6 +96,14 @@ def with_type(type)
       end
     end
 
+    def deep_dup # :nodoc:
```
Comment (matthewd) [inline activemodel/lib/active_model/attribute.rb:99]: Could we apply this optimization on `dup` in general? The caller is nearby, so it's hardly a layering violation, but it still feels just a little more implementation-aware than we could be, maybe?

(Incidentally notable: our actual `initialize_dup`'s value-dupping seems surprisingly casual -- likely doing so when unnecessary, possibly missing necessary depth in other cases?)

Comment (byroot) [reply activemodel/lib/active_model/attribute.rb:99]: > The caller is nearby, so it's hardly a layering violation, but it still feels just a little more implementation-aware than we could be, maybe?

Yeah, maybe we should just not call `dup` nor `deep_dup` but a more explicit method? something like `deep_dup_unless_shareable` (or something better that convey the same meaning)

Comment (byroot) [reply activemodel/lib/active_model/attribute.rb]: Look like that broke CI quite a bit 🤔

Comment (byroot) [reply activemodel/lib/active_model/attribute.rb]: Nevermind, the issue was that I changed `attributes` for `@attributes`, some subclass do redefine that accessor.

File: activemodel/lib/active_model/type/time.rb [hunk 2]
```diff
@@ -46,6 +46,14 @@ def type
         :time
       end
 
+      def mutable? # :nodoc:
+        # Time#zone can be mutated by #utc or #localtime
+        # However when serializing the time zone will always
+        # be coerced and even if the zone was mutated Time instances
+        # remain equal, so we don't need to implement `#changed_in_place?`
```
Comment (jhawthorn) [inline activemodel/lib/active_model/type/time.rb:53]: I think this makes sense 👍

**Review by jhawthorn — APPROVED** [review]

**Comment by george-ma** [comment]
had a concern about this PR Jean, if you have a sec to give your thoughts cc: @byroot https://github.com/Shopify/shopify/pull/569940#issuecomment-2605036320

**Pull request merged**